refactor: withDefaults and type props declaration in molecules

ArkadiuszKrzymienski opened the pull request:
# Related issue
Molecules part of #83 

# Scope of work
- molecules use withDefaults method and type-based declaration to define component props.
- better typing for inherited attributes and nested components.
- better typing for provide/inject

AlanLes commented on src/components/molecules/UiInteractiveSvg/_internal/UiInteractiveSvgElement.vue:
```diff
@@ -24,34 +24,35 @@ import {
+withDefaults(defineProps<InteractiveSvgElementProps>(), { tag: 'path' });
+const setAttrs: InteractiveSvgAttrsFunc | Ref<InteractiveSvgAttrsFunc> | undefined = inject('elementsAttrs');
+const elementsAttrs: InteractiveSvgAttrsFunc = (attrs) => {
+  const elementsAttrs = (setAttrs as Ref<InteractiveSvgAttrsFunc>).value || setAttrs;
```
Q: Is this cast necessary? What can we do to get rid of it, to avoid it? 🤔 🙂

ArkadiuszKrzymienski replied:
thanks, we can use `isRef` 😁

AlanLes commented on src/components/molecules/UiRating/UiRating.vue:
```diff
@@ -94,7 +94,7 @@
               v-bind="textLabelAttrs"
               class="visual-hidden"
-            >{{ defaultProps.translation.stars(index) }}</span>
+            >{{ defaultProps.translation!.stars(index) }}</span>
```
Q: What's this `!.` operator? 🤔 😲 😄

AlanLes commented on src/components/molecules/UiScale/UiScale.vue:
```diff
@@ -302,38 +288,36 @@ if (translationMild.value) {
   numberStepperAttrs: {
-    buttonDecrementAttrs: buttonDecrementAttrs.value,
-    buttonIncrementAttrs: buttonIncrementAttrs.value,
+    buttonDecrementAttrs: buttonDecrementAttrs?.value as ButtonAttrsProps,
```
Q: Is there maybe any way to get rid of this cast? 😄

AlanLes commented on src/components/molecules/UiStepper/_internal/UiStepperStep.vue:
```diff
@@ -37,35 +37,38 @@ export default { inheritAttrs: false };
-    type: Number,
-    default: 0,
-  },
+  indexOfActiveStep?: number;
```
Doesn't `activeStepIndex` sound better? 🤔 🙂

AlanLes commented on src/components/molecules/UiToggleButtonGroup/_internal/UiToggleButton.vue:
```diff
@@ -25,42 +25,45 @@ import {
+
+const props = withDefaults(defineProps<ToggleButtonProps>(), { value: '' });
+const emit = defineEmits<ToggleButtonEmits>();
+const attrs = useAttrs() as ToggleButtonAttrsProps;
```
I would try to get rid of this cast :)

AlanLes commented on src/components/molecules/UiToggleButtonGroup/_internal/UiToggleButton.vue:
```diff
@@ -25,42 +25,45 @@ import {
   }
 }
-const modelValue = inject('modelValue') as WritableComputedRef<ToggleButtonValue>;
+const modelValue = inject('modelValue') as WritableComputedRef<ToggleButtonGroupProps['modelValue']>;
```
Here as well

AlanLes commented on src/components/templates/UiQuestion/UiQuestion.vue:
```diff
@@ -299,7 +299,7 @@ const defaultProps = computed(() => ({
   },
   notificationFeedbackAttrs: {
-    type: 'success' as NotificationType,
+    type: 'success' as NotificationProps['type'],
```
IMO this cast is unnecessary 🤔 🙂

ArkadiuszKrzymienski replied:
I'll remove this in PR with templates components because then it won't be needed anymore. 👍

AlanLes reviewed (approved):
Looks fine, I would just recommend thinking more when we're using cast-mechanism (` as `) because, in 90% of cases, it's not what we should do, and THERE IS another way of doing so. 🙂 Besides that - great job, man! 🥳 🎉

ArkadiuszKrzymienski commented on .eslintrc.js:
```diff
@@ -12,12 +12,6 @@ module.exports = {
-  globals: {
-    defineProps: 'readonly',
-    defineEmits: 'readonly',
-    defineExpose: 'readonly',
```
Yes, eslint doesn't warn about it. From `vue-eslint-parser` v9.0.0 this is no longer needed. 😀

ArkadiuszKrzymienski commented on src/components/molecules/UiDropdown/_internal/UiDropdownItem.vue:
```diff
@@ -23,71 +23,70 @@ import {
-const changeHandler = inject('changeHandler') as (value: DropdownValue) => void;
-const dropdownItemKeydownHandler = inject('dropdownItemKeydownHandler') as (e: { key: string }) => Promise<void>;
-const modelValue = inject('modelValue') as ComputedRef<DropdownValue>;
+const changeHandler = inject('changeHandler') as (value: DropdownProps['modelValue']) => void;
```
Ok I will change the approach to `inject` and use it with generic, thanks for that 💪

ArkadiuszKrzymienski commented on src/components/molecules/UiDropdown/_internal/UiDropdownItem.vue:
```diff
@@ -23,71 +23,70 @@ import {
   },
 }));
-const attrs = useAttrs();
+const attrs = useAttrs() as DropdownItemAttrsProps;
```
As far as I know, `useAttrs` cannot be a generic function, always returns type `Record<string, unknown>`

AlanLes replied:
Ah, you're right; I thought it was our custom composable function, but it's indeed imported from the `'vue'`, sorry. In this scenario, I think we should build our custom `type-guards`, which we'll then use along with this `attrs` variable, just to make sure about the type of it. Because, again, people can pass any `attrs` they want, and we need to assure TypeScript (**with type-guard**) about types of these :) For now, I think it may stay as it is implemented right now, but I believe it should be refactored in every case to narrow the type using type-guard because it's a good pattern to use them. :)  Maybe let's create a separate issue out of that? :)

ArkadiuszKrzymienski replied:
I would also stay with the current implementation for now. We will release version 0.0.6 soon, which will drop backward compatibility for props and we will remove all `useAttrs` from components. IMO there is no need to create a new issue.  WDYT ?

AlanLes replied:
Sounds great! 👍 😄

ArkadiuszKrzymienski commented:
Main changes after the latest CR:
- improved typing for inject/provider
- improved typing for the `UiInteractiveSvg` component
- where possible, removed type assertions
- removed unnecessary types."

AlanLes commented on src/components/molecules/UiInteractiveSvg/_internal/UiInteractiveSvgElement.vue:
```diff
@@ -22,42 +22,47 @@ import {
   }
 }
-provide('elementsAttrs', nestedElementsAttrs);
+provide<Ref<InteractiveSvgFunc> | InteractiveSvgFunc>('elementsAttrs', nestedElementsAttrs);
```
Q: Why is this either `ref` or _not-ref_? 🤔

ArkadiuszKrzymienski replied:
In the case of nested elements in elements, we provide 'computed', which is why 'ref' is needed.  I will try to simplify it. 😉

AlanLes commented on src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:
```diff
@@ -21,49 +21,49 @@ import {
-const props = defineProps({
+export type LoaderTransitionState = 'from' | 'active' | 'to';
+export type LoaderTransitionMove = 'leave' | 'enter';
+export type LoaderTransitionHook = `before-${LoaderTransitionMove}`
```
Nice! It could be simplified even more, like
```typescript
export type LoaderTransitionHook = `${'before' | 'after'}-${LoaderTransitionMove}`
  | `${LoaderTransitionMove}`
  | `${LoaderTransitionMove}-cancelled`;
```

AlanLes commented on src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:
```diff
@@ -21,49 +21,49 @@ import {
 const contentEl = ref<HTMLElement | null>(null);
 const isLoaderVisible = ref(props.isLoading);
 const transitionEl = computed(() => (isLoaderVisible.value ? loaderEl.value : contentEl.value) as HTMLElement);
 const move = computed(() => (isLoaderVisible.value === props.isLoading ? 'enter' : 'leave'));
```
What do you think about creating const variables for these `'enter'` and `'leave'` and reusing them both here and in the type declaration? 
```typescript
const ENTER_MOVE = 'enter';
const LEAVE_MOVE = 'leave';
export type LoaderTransitionMove = typeof ENTER_MOVE | typeof LEAVE_MOVE;
...
const move = computed<LoaderTransitionMove>(() => (isLoaderVisible.value === props.isLoading ? ENTER_MOVE : LEAVE_MOVE));
```

ArkadiuszKrzymienski replied:
great idea 👍

AlanLes commented on src/types/tag.ts:
```diff
@@ -1,3 +1,3 @@
 export type HTMLTag = keyof HTMLElementTagNameMap;
 
-export type ListTag = 'ol'|'ul';
+export type ListHTMLTag = 'ol'|'ul';
```
Doesn't it sound better as `HTMLListTag`? Just an idea :D 🙈

ArkadiuszKrzymienski replied:
Yes, you're right, it sounds better 👍 😄

AlanLes reviewed (approved):
Review done. Great job man! A lot of work! You're on fire! 🔥 🚀 😄

ArkadiuszKrzymienski commented on src/components/atoms/UiRange/UiRange.vue:
```diff
@@ -127,50 +131,52 @@ const props = withDefaults(defineProps<RangeProps>(), {
+  ...attrs.value,
 }));
 </script>
 
```
Weird 👀 I only see new changes 😅 I don't know why 🙈

![range](https://user-images.githubusercontent.com/82519187/218446760-1a06b954-9fd3-4dc9-a44a-dce3db15ba07.png)

ArkadiuszKrzymienski commented on src/components/molecules/UiFormField/UiFormField.vue:
```diff
@@ -82,87 +82,71 @@ import {
   props.id || `input-${uid()}`
 ));
-// TODO: remove in 0.6.0 / BEGIN
+  // TODO: remove in 0.6.0 / BEGIN
```
Thanks, I added it by mistake 🙈

ArkadiuszKrzymienski commented on src/components/molecules/UiDropdown/_internal/UiDropdownItem.vue:
```diff
@@ -23,71 +23,74 @@ import {
-const dropdownItemKeydownHandler = inject('dropdownItemKeydownHandler') as (e: { key: string }) => Promise<void>;
-const modelValue = inject('modelValue') as ComputedRef<DropdownValue>;
+const changeHandler = inject<(value: Required<DropdownModelValue>) => void>('changeHandler');
+// eslint-disable-next-line @typescript-eslint/no-empty-function
```
Yes, you're right, `() => undefined` solves this eslint problem, thanks 🙂

ArkadiuszKrzymienski commented:
@AlanLes Many thanks for taking the time to review the code 🤝🚀😀
